### PR TITLE
bpo-35053: Enhance tracemalloc to trace free lists

### PR DESCRIPTION
**Pull request by vstinner**
tracemalloc now tries to update the traceback when an object is
reused from a "free list" (optimization for faster object creation,
used by the builtin list

Changes:

* Add _PyTraceMalloc_NewReference() function which tries to update
  the Python traceback of a Python object.
* _Py_NewReference() now calls _PyTraceMalloc_NewReference().

https://bugs.python.org/issue35053

### CODE REVIEW

**Comment by vstinner** [comment]
I modified my PR to make the overhead not significant:

```
Mean +- std dev: [ref] 20.8 ns +- 0.7 ns -> [patch] 20.5 ns +- 0.3 ns: 1.01x faster (-1%)
```

https://bugs.python.org/issue35053#msg328361

**Comment by vstinner** [comment]
I also optimized _PyTraceMalloc_NewReference() to make it faster when tracemalloc is tracing memory allocations.

**Comment by vstinner** [comment]
@methane: Ok, the PR is now waiting for your review :-)

**Comment by vstinner** [comment]
See also PR #10079 and https://bugs.python.org/issue35059 which convert the _Py_NewReference() macro to a "static inline" function.

**Review by methane — APPROVED** [review]

**Comment by vstinner** [comment]
@methane: Thanks for the review!

While reviewing one last time before merging, I found a potential issue: traceback_new() was called when tables were not locked, whereas all other calls to traceback_new() are done with tables locked. I moved the traceback_new() call inside TABLES_LOCK/TABLES_UNLOCK to make the code safer *and* faster (traceback_new is avoided if it's not needed, if the object is not tracked by tracemalloc).

**Pull request merged**
